src/portable/nxp/transdimension/dcd_transdimension.c

```diff
@@ -314,6 +316,39 @@ static void qtd_init(dcd_qtd_t* p_qtd, void * data_ptr, uint16_t total_bytes)
   }
 }
 
+static void qtd_init_fifo(dcd_qtd_t* p_qtd, tu_fifo_buffer_info_t *info, uint16_t total_bytes)
+{
+  tu_memclr(p_qtd, sizeof(dcd_qtd_t));
+
+  p_qtd->next        = QTD_NEXT_INVALID;
+  p_qtd->active      = 1;
+  p_qtd->total_bytes = p_qtd->expected_bytes = total_bytes;
+
+  // Fifo length has been trimmed to total_bytes
+  int16_t len_lin  = info->len_lin;
+
+  if (len_lin != 0)
+  {
+    p_qtd->buffer[0]   = (uint32_t) info->ptr_lin;
+
+    len_lin -= 4096 - ((uint32_t) info->ptr_lin - tu_align4k((uint32_t) info->ptr_lin));
+
+    // Set linear part
+    uint8_t i = 1;
+    for(; i<5; i++)
+    {
+      if (len_lin <= 0) break;
+      p_qtd->buffer[i] |= tu_align4k( p_qtd->buffer[i-1] ) + 4096;
+      len_lin -= 4096;
+    }
+    // Set wrapped part
+    for(uint8_t page = 0; i<5; i++, page++)
+    {
+      p_qtd->buffer[i] |= (uint32_t) info->ptr_wrap + 4096 * page;
+    }
```

Comment:
I don't quite understand these lines, are you trying to queue both linear and wrap buffer into one TD ?  Which I don't think is possible, we probably need 2 TDs for this.

Reply:
Yes I'm trying to queue both parts in 5*4kB TDs. Fifo need to be 4k aligned and has a size of multiple of 4K, for a max of 20K.
TD[0] can start at a arbitrary address and TD[1]-TD[4] must start at 4k boundary.

Take an example of a 8k fifo filled between 2000-100:
![image](https://user-images.githubusercontent.com/4375114/132499752-803c736c-9641-4e80-897d-879ed2cd2608.png)

TD[0] is set to offset of `fifo->ptr_lin`, then each TD is aligned to 4k until len_lin is finished.

Then it wrap to top and start at offset 0.

Reply:
I think you mis-understand the TD specs, within this function, it only preps 1 TD. The buffer[] page array is only used when the total bytes is large enough and memory cross the 4k boundary e.g when total bytes is 20 KB.

Reply:
> within this function, it only preps 1 TD

You're right. 

In fact I mean 5 pointers within the same TD, sorry for the confusion.
![image](https://user-images.githubusercontent.com/4375114/132520378-7c7abb95-eb2f-4f87-a805-b34eba08f210.png)

> we probably need 2 TDs for this.

I've tried 2 TDs, one for linear and one for wrapped part. In this case it cut the data into 2 separate transfers, for example cdc will receive half data.

Reply:
we can transfer 1 TD after another instead of using 2 TDs at once. Can you update the above code to initialize the buffer page correctly. It should only affected by ptr_lin, and is mostly identical to existing qtd_init(). In fact, I think we should just use the qtd_init(ptr_lin, len) instead of having additional qtd_init_fifo()

Reply:
It won't work when you need continuous data, for example in UAC2.

And host will think the transfer is finished since len_lin can be a short length packet.

At the beginning I've already tried using 2 TDs since it's easier to implement, doesn't work at all.

Reply:
The current TD buffer setup is not correct in general, the requirement with 4K aligned is too specific to make its way into the buffer page array set up. And this does not work with other MCU dcd. 

For ISO, you have to understand this limit and need to declare the fifo depth to be multiple of ISO frame. Re-config the fifo length to even smaller depth than its buffer when needed when the ISO packet size is changed is better implementation. E.g fifo buffer is 4x1024, but should be changed to 40x100 when iso packet changed from 1024 to 100.

Reply:
> the requirement with 4K aligned is too specific to make its way into the buffer page array set up. And this does not work with other MCU dcd.

Yes it is specific, so I've introduced a on/off switch in #1063 to not enable fifo_xfer by default, except explicitly.

> For ISO, you have to understand this limit and need to declare the fifo depth to be multiple of ISO frame. 

That's won't work since sample length could be +-1, wrapped part has to be always considered.

Reply:
> Yes it is specific, so I've introduced a on/off switch in #1063 to not enable fifo_xfer by default, except explicitly.

I don't like the `TUD_AUDIO_PREFER_RING_BUFFER`, it is only temporarily solution, we will eventual remove that.

> That's won't work since sample length could be +-1, wrapped part has to be always considered.

Apparently, I am not aware of this issue. In this case, can you modify to make it more generic. xfer_fifo() will make use of qtd_init() with the linear part first. Then update the buffer pagep[] with wrapped part if following conditions is met
1. Linear len < total _bytes
2. wrap pointer is 4k alignment.

Reply:
> Then update the buffer pagep[] with wrapped

you mean modify `buffer pagep[]` inside ` xfer_fifo()` and remove `qtd_init_fifo()` ?

Reply:
it is mostly to reduce code duplication, for now, just keep it as it is if it is too confused. Only set up the buffer page only above 2 condition are met. currently wrapped part is always set up.

Reply:
@HiFiPhile to clarify: not all edpt_xfer_fifo() need to combine linear + wrapped part, e.g CDC should have no issue breaking these into 2 sub transfers (one after another when USB complete ISR). Therefore not all fifo buffer need to be 4K. It is mostly for ISO applcation/driver, so having a 4K check would fit both.